lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -3506,24 +3531,66 @@ FIRCircuitParser::parseCircuit(const llvm::MemoryBuffer *annotationsBuf) {
 
 DoneParsing:
 
+  auto reduceTargets =
+      [](const SmallVector<StringAttr> &acc,
+         SmallVector<StringAttr> x) -> SmallVector<StringAttr> {
+    if (x.empty())
+      return acc;
+    // Collisions are impossible, just append.
+    SmallVector<StringAttr> retval(acc);
+    retval.append(x.begin(), x.end());
+    return retval;
+  };
+
   // Now that we've parsed all the prototypes and created all the module ops,
   // go ahead and parse all their bodies.  This can be done in parallel.
   if (getContext()->isMultithreadingEnabled()) {
     mlir::ParallelDiagnosticHandler diagHandler(getContext());
     std::atomic<bool> anyFailed{false};
-    llvm::parallelForEachN(0, deferredModules.size(), [&](size_t index) {
-      diagHandler.setOrderIDForThread(index);
-      if (parseModuleBody(deferredModules[index]))
-        anyFailed = true;
-    });
+    std::atomic<unsigned> threadID{0};
```

Comment:
Random FYI, but MLIR now has mlir/IR/Threading.h which we should adopt to clean up our threading stuff.  IT handles the diagnostics switch and the "if mlir paralllelism enabled" thing.  It doesn't have reduce though, it would be great to upstream that if you need it.

Reply:
This was perhaps too cute.  I changed this to just do the merge in a loop after this.  I do agree that porting the `parallelTransformReduce` method to the new threading API would be great (and it would be good for me to actually do some upstream MLIR work...).  That method is used in `LowerToHW`.

Reply:
I can take care of changing to `mlir/IR/Threading.h` in a separate, unrelated commit.